server/mailet/mailets/src/main/java/org/apache/james/transport/mailets/RecipientRewriteTableProcessor.java

```diff
@@ -0,0 +1,313 @@
+        @Override
+        public MailAddress apply(Optional<MailAddress> mailAddress) {
+            return mailAddress.get();
+        }
+    };
+
+    private static final Predicate<Optional<MailAddress>> mailAddressPresent = new Predicate<Optional<MailAddress>>() {
+        @Override
+        public boolean apply(Optional<MailAddress> mailAddress) {
+            return mailAddress.isPresent();
+        }
+    };
+
+    private static final Function<Mapping, Optional<MailAddress>> mailAddressFromMapping = new Function<Mapping, Optional<MailAddress>>() {
+        @Override
+        public Optional<MailAddress> apply(Mapping addressMapping) {
+            try {
+                return Optional.of(new MailAddress(addressMapping.asString()));
+            } catch (AddressException e) {
+                return Optional.absent();
+            }
+        }
+    };
+
+    public RecipientRewriteTableProcessor(RecipientRewriteTable virtualTableStore, DomainList domainList) {
+        this.virtualTableStore = virtualTableStore;
+        this.domainList = domainList;
+    }
+
+    public void processMail(Mail mail, MailetContext mailetContext) throws MessagingException{
+        ImmutableList<RrtExecutionResult> mappingDatas = toMappingDatas(mail, mailetContext);
+
+        ImmutableList<MailAddress> newRecipients = getRecipientsByCondition(mappingDatas, recipientWithoutError);
+
+        ImmutableList<MailAddress> errorMailAddresses = getRecipientsByCondition(mappingDatas, recipientWithError);
+
+        if (!errorMailAddresses.isEmpty()) {
+            mailetContext.sendMail(mail.getSender(), errorMailAddresses, mail.getMessage(), Mail.ERROR);
+        }
+
+        if (newRecipients.isEmpty()) {
+            mail.setState(Mail.GHOST);
+        }
+
+        mail.setRecipients(newRecipients);
+    }
+
+    private ImmutableList<MailAddress> getRecipientsByCondition(ImmutableList<RrtExecutionResult> mappingDatas, Predicate<RrtExecutionResult> filterCondition) {
+        return FluentIterable.from(mappingDatas)
+            .filter(filterCondition)
+            .transformAndConcat(mailAddressesFromMappingData)
+            .toList();
+    }
+
+    private ImmutableList<RrtExecutionResult> toMappingDatas(final Mail mail, final MailetContext mailetContext) {
+        Function<MailAddress, RrtExecutionResult> convertToMappingData = new Function<MailAddress, RrtExecutionResult>() {
+            @Override
+            public RrtExecutionResult apply(MailAddress recipient) {
+                Preconditions.checkNotNull(recipient);
+
+                return getRrtExecutionResult(mail, mailetContext, recipient);
+            }
+
+        };
+
+        return FluentIterable.from(mail.getRecipients())
+            .transform(convertToMappingData)
+            .toList();
+    }
+
+    private RrtExecutionResult getRrtExecutionResult(Mail mail, MailetContext mailetContext, MailAddress recipient) {
```

Comment:
MailetContext should be a field of this class ^^ 

It will avoid you passing it all along ;-)